sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Listeners/EventHubListener.cs

```diff
@@ -0,0 +1,377 @@
+
+                return Task.CompletedTask;
+            }
+
+            public async Task ProcessEventsAsync(ProcessorPartitionContext context, IEnumerable<EventData> messages)
+            {
+                var triggerInput = new EventHubTriggerInput
+                {
+                    Events = messages.ToArray(),
+                    PartitionContext = context
+                };
+
+                TriggeredFunctionData input = null;
+                if (_singleDispatch)
+                {
+                    // Single dispatch
+                    int eventCount = triggerInput.Events.Length;
+                    List<Task> invocationTasks = new List<Task>();
+                    for (int i = 0; i < eventCount; i++)
+                    {
+                        if (_cts.IsCancellationRequested)
+                        {
+                            break;
+                        }
+
+                        EventHubTriggerInput eventHubTriggerInput = triggerInput.GetSingleEventTriggerInput(i);
+                        input = new TriggeredFunctionData
+                        {
+                            TriggerValue = eventHubTriggerInput,
+                            TriggerDetails = eventHubTriggerInput.GetTriggerDetails(context)
+                        };
+
+                        Task task = TryExecuteWithLoggingAsync(input, triggerInput.Events[i]);
+                        invocationTasks.Add(task);
+                    }
+
+                    // Drain the whole batch before taking more work
+                    if (invocationTasks.Count > 0)
+                    {
+                        await Task.WhenAll(invocationTasks).ConfigureAwait(false);
+                    }
+                }
+                else
+                {
+                    // Batch dispatch
+                    input = new TriggeredFunctionData
+                    {
+                        TriggerValue = triggerInput,
+                        TriggerDetails = triggerInput.GetTriggerDetails(context)
+                    };
+
+                    using (_logger.BeginScope(GetLinksScope(triggerInput.Events)))
+                    {
+                        await _executor.TryExecuteAsync(input, _cts.Token).ConfigureAwait(false);
+                    }
+                }
+
+                // Checkpoint if we processed any events.
+                // Don't checkpoint if no events. This can reset the sequence counter to 0.
+                // Note: we intentionally checkpoint the batch regardless of function
+                // success/failure. EventHub doesn't support any sort "poison event" model,
+                // so that is the responsibility of the user's function currently. E.g.
+                // the function should have try/catch handling around all event processing
+                // code, and capture/log/persist failed events, since they won't be retried.
+                if (messages.Any())
```

Comment:
nit:  This is potentially enumerating the sequence a second time.   You've got an array of these in the `triggerInput` that you may want to just read the length of.